### PR TITLE
Implement BigtableAsyncTableRegionLocator

### PR DESCRIPTION
**Pull request by rupeshitpatekarq**
Implemented getRegionLocation() method in BigtableAsyncTableRegionLocator class & created AbstractBigtbleRegionLocator to share code between BigtableAsyncTableRegionLocator and BigtableRegionLocator classes. Could you please review implementation and confirm.

### CODE REVIEW

**Comment by sduskis** [comment]
Overall, the Async Region Locator needs to be actually async. I'll be happy to help you figure this out further.

**Comment by rupeshitpatekarq** [comment]
@sduskis , Implemented getRegionLocatidateon() method in async way & added test for the same. Verified against all tests in both mode. Also done with earlier code review changes. Could you please review it.

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTableRegionLocator.java [hunk 1]
```diff
@@ -64,7 +57,45 @@ public TableName getName() {
...
 
   @Override
-  public CompletableFuture<HRegionLocation> getRegionLocation(byte[] arg0, boolean arg1) {
-    throw new UnsupportedOperationException("getRegionLocation"); // TODO
+  public CompletableFuture<HRegionLocation> getRegionLocation(byte[] row, boolean reload){
+    try {
+      this.regionsFuture = getRegionsAsync(reload);
+    } catch (IOException e) {
+      //As getRegionLocation() method is not throwing any exception hence needs to handle following way.
+      e.getCause();
```
Comment (sduskis) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTableRegionLocator.java]: Please `return FutureUtils.failedFuture(e)` here.  `getRegionsAsync()` does have some logic to throw an IOException.  Or, even better,  `getRegionsAsync()` should not throw an exception, and should do its own wrapping of the execution with a `Futures.immediateFailFuture()`

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTableRegionLocator.java [hunk 2]
```diff
@@ -64,7 +57,45 @@ public TableName getName() {
...
+    }
+    Futures.addCallback(this.regionsFuture, new FutureCallback<List<HRegionLocation>>() {
+      @Override 
+      public void onSuccess(@Nullable List<HRegionLocation> result) {
+        for(HRegionLocation region : result) {
+          if (region.getRegion().containsRow(row)) {
+            hRegionLocation = region;
+          }
```
Comment (sduskis) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTableRegionLocator.java]: pleas `break` in the if statement.

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTableRegionLocator.java [hunk 3]
```diff
@@ -64,7 +57,45 @@ public TableName getName() {
...
+      }
+      @Override 
+      public void onFailure(Throwable t) {
+        synchronized (BigtableAsyncTableRegionLocator.this) {
+          regionsFuture = null;
+        }
+      }
+    });
+    if(hRegionLocation == null) {
+      LOG.info("Region not found for row: ", Bytes.toStringBinary(row));
```
Comment (sduskis) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTableRegionLocator.java]: this should probably be an exception.  This should never happen.

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTableRegionLocator.java [hunk 4]
```diff
@@ -64,7 +57,45 @@ public TableName getName() {
...
+    }
+    return CompletableFuture.completedFuture(hRegionLocation);
```
Comment (sduskis) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTableRegionLocator.java]: This is wrong.  The end result is still a synchronous call.

File: bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtbleRegionLocator.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+ */
+public abstract class AbstractBigtbleRegionLocator {
+  // Reuse the results from previous calls during this time.
+  /** Constant <code>MAX_REGION_AGE_MILLIS=60 * 1000</code> */
+  public static long MAX_REGION_AGE_MILLIS = 60 * 1000;
+
+  /** Constant <code>LOG</code> */
+  protected static final Logger LOG = new Logger(AbstractBigtbleRegionLocator.class);
+
+  protected final TableName tableName;
```
[the remark this answers is not in the document]
Comment (sduskis) [reply bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtbleRegionLocator.java]: please fix this.

[the remark this answers is not in the document]
Comment (rupeshitpatekarq) [reply bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtbleRegionLocator.java]: tableName & regionsFuture we are using in children (BigtableRegionLocator & BigtableAsyncTableRegionLocator) hence they are protected. LOG is made now as private.

**Comment by rupeshitpatekarq** [comment]
@sduskis , Done. Only tableName is protected as we are using it in its children (BigtableAsyncTableRegionLocator, & BigtableRegionLocator). Please review now.

**Comment by sduskis** [comment]
I'm going to merge this, and make the changes I want, which are on a small subset of the code.

**Pull request merged**